### PR TITLE
Permissions! :)

### PR DESCRIPTION
**Pull request by vegajdr**
image

.
.
.
.
.

Now the useful stuff:

#### Resource Authorization

Each resource will have an `authorization` class, similar to the concept of a `policy`
used in the `pundit` gem.

Inside this class you can define classes that correspond to the name of roles expected on an user, for example:

```ruby
class PostAuthorization < ResourceAuthorization
  class Author < RoleAbility
    # required, otherwise raises NotImplementedError
    def abilities
      can :create, Post
    end
  end
end
```
will automatically grant the `create` permission for a `Post` object on a signed in user
that has the `author` role associated to it

#### Guest Ability

If we want to allow access to a resource to a visitor that does not possess a user account, we can define
a `guest` ability the following way:

```ruby
class PostAuthorization < ResourceAuthorization
  # Note the inheritance of GuestAbility, this ensures the permissions
  # defined on this class get loaded first as base permissions, also does not have
  # a need to have `user` passed in as an argument
  class Guest < GuestAbility
    def abilities
      can :read, Post
    end
  end

  class Author < RoleAbility
    def abilities
      can :create, Post
    end
  end
end
```

This definition essentially opens the ability to read any `Post` even if you are not signed in.
The class is required to be named `Guest`

### CODE REVIEW

**Comment by ernbrn** [comment]
@vegajdr I'll add it to the Trello, but we'll need a concept of a Guest user that probably view most things 

https://stackoverflow.com/questions/5158326/safest-and-railsiest-way-in-cancan-to-do-guest-user-admin-permissions#answer-5417527

**Comment by vegajdr** [comment]
> @vegajdr I'll add it to the Trello, but we'll need Guest user permissions that can probably view most things
> 
> https://stackoverflow.com/questions/5158326/safest-and-railsiest-way-in-cancan-to-do-guest-user-admin-permissions#answer-5417527

@ernbrn This is how it's going to look like:
```ruby
# frozen_string_literal: true

class CreatorAuthorization < ResourceAuthorization
  class Guest < GuestAbility
    def abilities
      can :read, Creator
    end
  end

  class Admin < RoleAbility
    def abilities
      can :manage, Creator
    end
  end

  class Moderator < RoleAbility
    def abilities
      can :destroy, Creator
    end
  end
end
```

File: app/authorization/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+If we want to allow access to a resource to a visitor that does not possess a user account, we can define
+a `guest` ability the following way:
+
+```ruby
+class PostAuthorization < ResourceAuthorization
+  # Note the inheritance of GuestAbility, this ensures the permissions
+  # defined on this class get loaded first as base permissions, also does not have
+  # a need to have `user` passed in as an argument
+  class Guest < GuestAbility
```
Comment (ernbrn) [inline app/authorization/README.md]: 🙌 Love this

File: app/authorization/resource_authorization.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    return unless guest_permissions
+
+    merge(guest_permissions.new)
+  end
+
+  def authorization
+    return if user.blank?
+
+    user.roles.each do |role|
+      role_class = "#{self.class}::_#{role.name}".classify.safe_constantize
```
Comment (ernbrn) [inline app/authorization/resource_authorization.rb]: Neat!

File: app/controllers/concerns/authorized_resource.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    controller_name.singularize
+  end
+
+  def resource
+    instance_variable_get(:"@#{resource_name}")
+  end
+
+  def composed_ability
+    Ability.new(current_user).tap do |ability|
+      self.class::ABILITIES.each { |ability_class| ability.merge(ability_class.new(current_user)) }
```
Comment (ernbrn) [inline app/controllers/concerns/authorized_resource.rb]: What will this look like in the classes?

**Pull request merged**